src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -461,96 +556,106 @@ public void OnResponseData(ReadOnlySpan<byte> buffer, bool endStream)
             }
 
             // Determine if we have enough data to process up to complete final response headers.
             private (bool wait, bool isEmptyResponse) TryEnsureHeaders()
             {
                 lock (SyncObject)
                 {
-                    CheckIfDisposedOrAborted();
+                    CheckResponseBodyState();
 
-                    if (_state == StreamState.ExpectingHeaders || _state == StreamState.ExpectingIgnoredHeaders || _state == StreamState.ExpectingStatus)
+                    if (_responseProtocolState == ResponseProtocolState.ExpectingHeaders || _responseProtocolState == ResponseProtocolState.ExpectingIgnoredHeaders || _responseProtocolState == ResponseProtocolState.ExpectingStatus)
                     {
                         Debug.Assert(!_hasWaiter);
                         _hasWaiter = true;
                         _waitSource.Reset();
                         return (true, false);
                     }
-                    else if (_state == StreamState.ExpectingData || _state == StreamState.ExpectingTrailingHeaders)
+                    else if (_responseProtocolState == ResponseProtocolState.ExpectingData || _responseProtocolState == ResponseProtocolState.ExpectingTrailingHeaders)
                     {
                         return (false, false);
                     }
                     else
                     {
-                        Debug.Assert(_state == StreamState.Complete);
+                        Debug.Assert(_responseProtocolState == ResponseProtocolState.Complete);
                         return (false, _responseBuffer.ActiveLength == 0);
                     }
                 }
             }
 
             public async Task ReadResponseHeadersAsync(CancellationToken cancellationToken)
             {
-                // Wait for response headers to be read.
                 bool emptyResponse;
-                bool wait;
-
-                // Process all informational responses if any and wait for final status.
-                (wait, emptyResponse) = TryEnsureHeaders();
-                if (wait)
+                try
                 {
-                    await GetWaiterTask(cancellationToken).ConfigureAwait(false);
+                    // Wait for response headers to be read.
+                    bool wait;
 
+                    // Process all informational responses if any and wait for final status.
                     (wait, emptyResponse) = TryEnsureHeaders();
-                    Debug.Assert(!wait);
+                    if (wait)
+                    {
+                        await GetWaiterTask(cancellationToken).ConfigureAwait(false);
+
+                        (wait, emptyResponse) = TryEnsureHeaders();
+                        Debug.Assert(!wait);
+                    }
+                }
+                catch
+                {
+                    CancelRequestBody();
+                    CancelResponseBody();
```

Comment:
Same question (CancelResponseBody calls CancelRequestBody; do we need both here?)